### PR TITLE
Wrapping namespace Reduction in namespace at (#26606)

### PR DESCRIPTION
**Pull request by divyanshsinghvi**
1) Wrapped namespace `Reduction` in namespace `at`
2) Prefixed `at::` wherever `Reduction::` is used

### CODE REVIEW

File: aten/src/ATen/core/Reduction.h [hunk 1]
```diff
@@ -11,4 +12,5 @@ enum Reduction {
   Sum,              // Sum losses
   END
 };
-}
+} // namespace Reduction
+} // namespace at
```
Comment (yf225) [inline aten/src/ATen/core/Reduction.h]: @li-roy Do you remember if there a specific reason we don't wrap `Reduction` in `at::` in the original design?

Comment (li-roy) [reply aten/src/ATen/core/Reduction.h]: No I don't think so.

**Review by li-roy — APPROVED** [review]

File: test/test_cpp_api_parity.py [hunk 2]
```diff
@@ -218,15 +218,15 @@ def _python_arg_to_cpp_arg(self, python_arg):
...
-                    cpp_arg = 'Reduction::None'
+                    cpp_arg = 'at::Reduction::None'
                 elif python_arg == 'mean':
-                    cpp_arg = 'Reduction::Mean'
+                    cpp_arg = 'at::Reduction::Mean'
                 elif python_arg == 'sum':
-                    cpp_arg = 'Reduction::Sum'
-                return CppArg(type='Reduction::Reduction', value='{}'.format(cpp_arg))
+                    cpp_arg = 'at::Reduction::Sum'
+                return CppArg(type='at::Reduction::Reduction', value='{}'.format(cpp_arg))
```
Comment (yf225) [inline test/test_cpp_api_parity.py]: For C++ frontend we encourage using `torch::` namespace instead of `at::`, and it would be great to change all of the use sites here to `torch::`.

File: torch/csrc/api/include/torch/nn/options/loss.h [hunk 3]
```diff
@@ -24,7 +24,7 @@ struct TORCH_API HingeEmbeddingLossOptions {
   /// reach in order to incur zero loss. Default: 1
   TORCH_ARG(double, margin) = 1.0;
   /// Specifies the reduction to apply to the output. Default: Mean
-  TORCH_ARG(Reduction::Reduction, reduction) = Reduction::Mean;
+  TORCH_ARG(at::Reduction::Reduction, reduction) = at::Reduction::Mean;
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/options/loss.h]: Ditto in this file regarding changing `at::` to `torch::`.

**Review by yf225 — CHANGES_REQUESTED** [review]
Thanks so much for the fix @divyanshsinghvi! I left some very minor comments regarding using `torch::` in C++ frontend.

**Comment by yf225** [comment]
@ailzhang XLA build seems to fail with
```
Oct 05 20:02:56 torch_xla/csrc/tensor_ops.cpp:74:23: error: use of undeclared identifier 'Reduction'; did you mean 'at::Reduction'?
Oct 05 20:02:56   return reduction == Reduction::Mean
Oct 05 20:02:56                       ^~~~~~~~~
Oct 05 20:02:56                       at::Reduction
Oct 05 20:02:56 /opt/conda/lib/python3.6/site-packages/torch/include/ATen/core/Reduction.h:4:11: note: 'at::Reduction' declared here
Oct 05 20:02:56 namespace Reduction {
Oct 05 20:02:56           ^
Oct 05 20:02:56 torch_xla/csrc/tensor_ops.cpp:107:10: error: use of undeclared identifier 'Reduction'; did you mean 'at::Reduction'?
Oct 05 20:02:56     case Reduction::None:
Oct 05 20:02:56          ^~~~~~~~~
Oct 05 20:02:56          at::Reduction
Oct 05 20:02:56 /opt/conda/lib/python3.6/site-packages/torch/include/ATen/core/Reduction.h:4:11: note: 'at::Reduction' declared here
Oct 05 20:02:56 namespace Reduction {
Oct 05 20:02:56           ^
Oct 05 20:02:56 torch_xla/csrc/tensor_ops.cpp:109:10: error: use of undeclared identifier 'Reduction'; did you mean 'at::Reduction'?
Oct 05 20:02:56     case Reduction::Mean:
Oct 05 20:02:56          ^~~~~~~~~
Oct 05 20:02:56          at::Reduction
Oct 05 20:02:56 /opt/conda/lib/python3.6/site-packages/torch/include/ATen/core/Reduction.h:4:11: note: 'at::Reduction' declared here
Oct 05 20:02:56 namespace Reduction {
Oct 05 20:02:56           ^
Oct 05 20:02:56 torch_xla/csrc/tensor_ops.cpp:112:10: error: use of undeclared identifier 'Reduction'; did you mean 'at::Reduction'?
Oct 05 20:02:56     case Reduction::Sum:
Oct 05 20:02:56          ^~~~~~~~~
Oct 05 20:02:56          at::Reduction
Oct 05 20:02:56 /opt/conda/lib/python3.6/site-packages/torch/include/ATen/core/Reduction.h:4:11: note: 'at::Reduction' declared here
Oct 05 20:02:56 namespace Reduction {
Oct 05 20:02:56           ^
Oct 05 20:02:56 torch_xla/csrc/tensor_ops.cpp:142:10: error: use of undeclared identifier 'Reduction'; did you mean 'at::Reduction'?
Oct 05 20:02:56     case Reduction::None:
Oct 05 20:02:56          ^~~~~~~~~
Oct 05 20:02:56          at::Reduction
Oct 05 20:02:56 /opt/conda/lib/python3.6/site-packages/torch/include/ATen/core/Reduction.h:4:11: note: 'at::Reduction' declared here
Oct 05 20:02:56 namespace Reduction {
Oct 05 20:02:56           ^
Oct 05 20:02:56 torch_xla/csrc/tensor_ops.cpp:143:10: error: use of undeclared identifier 'Reduction'; did you mean 'at::Reduction'?
Oct 05 20:02:56     case Reduction::Sum:
Oct 05 20:02:56          ^~~~~~~~~
Oct 05 20:02:56          at::Reduction
Oct 05 20:02:57 /opt/conda/lib/python3.6/site-packages/torch/include/ATen/core/Reduction.h:4:11: note: 'at::Reduction' declared here
Oct 05 20:02:57 namespace Reduction {
Oct 05 20:02:57           ^
Oct 05 20:02:57 torch_xla/csrc/tensor_ops.cpp:145:10: error: use of undeclared identifier 'Reduction'; did you mean 'at::Reduction'?
Oct 05 20:02:57     case Reduction::Mean: {
Oct 05 20:02:57          ^~~~~~~~~
Oct 05 20:02:57          at::Reduction
Oct 05 20:02:57 /opt/conda/lib/python3.6/site-packages/torch/include/ATen/core/Reduction.h:4:11: note: 'at::Reduction' declared here
Oct 05 20:02:57 namespace Reduction {
Oct 05 20:02:57           ^
Oct 05 20:03:02 7 errors generated.
```
because we are moving `Reduction` into `at::` namespace. Do you think we should issue a quick fix to XLA after merging this PR, or should we have a bridging PR that alias `at::Reduction` to `Reduction` first, and remove the alias after XLA is migrated?

**Comment by yf225** [comment]
NOTE: if https://github.com/pytorch/pytorch/pull/27345 is landed first, we will need to change `Reduction` usage in `torch::nn::CosineEmbeddingLoss` as well.

File: torch/csrc/api/include/torch/nn/options/loss.h (hunk 3 above)
Comment (divyanshsinghvi) [reply torch/csrc/api/include/torch/nn/options/loss.h]: I may be totally wrong in my understanding. But I was amazed at how the code would work with `torch::` instead of `at::`. So, I tried to dig deeper and found the types.h class which wrap around the object with `torch` namespace and with `using namespace at` we are essentially removing the need of use of `at::`. Please do correct me if my understanding is wrong.

Comment (yf225) [reply torch/csrc/api/include/torch/nn/options/loss.h]: Your understanding is 100% correct :D

**Comment by yf225** [comment]
@divyanshsinghvi https://github.com/pytorch/pytorch/pull/27345 is recently landed which has another occurrence of `Reduction` (in torch/csrc/api/include/torch/nn/options/loss.h), and we might need to change it as well. Thanks so much for your help!

**Comment by divyanshsinghvi** [comment]
> @divyanshsinghvi #27345 is recently landed which has another occurrence of `Reduction` (in `torch/csrc/api/include/torch/nn/options/loss.h`), and we might need to change it as well. Thanks so much for your help!

Yes, this PR includes changes for this.

**Comment by yf225** [comment]
@divyanshsinghvi Awesome thanks a lot! After CI passes I will work with @ailzhang to land this PR and also fix XLA build at the same time. :)

**Comment by divyanshsinghvi** [comment]
@yf225 I think the two fails are now because of xla. According to me, these three files would require the changes in xla repository. 
        test/cpp/test_aten_xla_tensor.cpp
        torch_xla/csrc/aten_xla_type.cpp
        torch_xla/csrc/tensor_ops.cpp

**Comment by yf225** [comment]
@divyanshsinghvi Yes I think so - I just opened a PR to the XLA repo: https://github.com/pytorch/xla/pull/1176, and we will merge that immediately after merging this PR to unbreak XLA build.

**Review by yf225 — APPROVED** [review]
@divyanshsinghvi Thanks so much for the awesome work!

**Comment by yf225** [comment]
I will merge this PR after internal CI passes.

File: test/cpp/api/modules.cpp [hunk 4]
```diff
@@ -1097,7 +1097,7 @@ TEST_F(ModulesTest, MultiLabelSoftMarginLossWeightedNoReduction) {
   auto input = torch::tensor({{0., 2., 2., 0.}, {2., 1., 0., 1.}}, torch::requires_grad());
   auto target = torch::tensor({{0., 0., 1., 0.}, {1., 0., 1., 1.}}, torch::kFloat);
   auto weight = torch::tensor({0.1, 0.6, 0.4, 0.8}, torch::kFloat);
-  auto options = MultiLabelSoftMarginLossOptions().reduction(Reduction::None).weight(weight);
+  auto options = MultiLabelSoftMarginLossOptions().reduction(at::Reduction::None).weight(weight);
```
Comment (divyanshsinghvi) [inline test/cpp/api/modules.cpp]: @yf225  why is it `at::` instead of `torch::`

Comment (divyanshsinghvi) [inline test/cpp/api/modules.cpp]: @yf225  why is it `at::` instead of `toorch::`

Comment (yf225) [reply test/cpp/api/modules.cpp]: @divyanshsinghvi thanks a lot for the catch! I changed it to `torch::`

**Comment by yf225** [comment]
@divyanshsinghvi It seems that some unrelated commits were accidentally added into this PR branch, so I cleaned them up and rebase the PR on top of current master.

**Comment by divyanshsinghvi** [comment]
> @divyanshsinghvi It seems that some unrelated commits were accidentally added into this PR branch, so I cleaned them up and rebase the PR on top of current master.

Yes, thanks for cleaning them up.
